### DISCUSSION TITLE
RequestExecutor using Webclient

### DISCUSSION THREAD

**Discussion by rmishra3261** [Netflix/dgs-framework#668] [Q&A] [answered]
I need an example of RequestExecutor which uses Webclient. I am using below code but it does not work:

```
    private RequestExecutor createWebClientRequestExecutor(HttpHeaders requestHeaders){
        return (url, headers, body) -> {
            
            Mono<String> response = webClient.post().uri(url).
                    header("Accept","application/json")
                    .header("Content-Type","application/json").bodyValue(body)
                    .retrieve().bodyToMono(String.class);

            return new HttpResponse(200,response.block(Duration.ofSeconds(2L)));
        };
    }
```
Please advise.

**Comment by paulbakker** [comment, marked as answer]
We just added an out-of-the-box WebClient based client: https://github.com/Netflix/dgs-framework/releases/tag/v4.9.0-rc.1

**Reply by rmishra3261** [reply]
Thanks a lot @paulbakker. I will try it out. 
Quick question: Is it also possible to use Webclient with CustomGraphQLClient or we have to use CustomMonoGraphQLClient?

**Reply by paulbakker** [reply]
It could be either. If you put a `block()` at the end of the `WebClient` call it unwraps the mono.

**Reply by rmishra3261** [reply]
Thanks Paul. Can you please also give me an example of a lambda body for MonoRequestExecutor.  I used below code but I am not getting any results:

```
public static Mono<GraphQLResponse> executeMonoQuery(String gqlEndpoint, String generatedQuery, MonoRequestExecutor monoRequestExecutor){

        MonoGraphQLClient monoGraphQLClient = new DefaultGraphQLClient(gqlEndpoint);
        return monoGraphQLClient.reactiveExecuteQuery(generatedQuery,new HashMap<>(),monoRequestExecutor);
}
```

Here is my webclient and MonoRequestExecutor:

 ```
private WebClient webClient = WebClient.builder().defaultHeaders(httpHeaders1 -> httpHeaders1.putAll(httpHeaders)).build();
    private MonoRequestExecutor createWebClientRequestExecutor(HttpHeaders requestHeaders){
        return (url, headers, body) -> {

            return webClient.post().uri(url).
                    header("Accept","application/json")
                    .header("Content-Type","application/json").bodyValue(body)
                    .retrieve().bodyToMono(HttpResponse.class);
        };
    }
```

**Reply by paulbakker** [reply]
Sure, that's possible and is how we used to integrate with WebClient before we added a WebClient specific implementation.
Basically you can use the same code we use in the WebClient implementation inside a `MonoRequestExecutor`: https://github.com/Netflix/dgs-framework/blob/master/graphql-dgs-client/src/main/kotlin/com/netflix/graphql/dgs/client/WebClientGraphQLClient.kt#L79

**Reply by paulbakker** [reply]
I would recommend justing the WebClientGraphQLClient implementation though. Why do you want to use it through a `Custom*` implementation?

**Reply by rmishra3261** [reply]
> Sure, that's possible and is how we used to integrate with WebClient before we added a WebClient specific implementation. Basically you can use the same code we use in the WebClient implementation inside a `MonoRequestExecutor`: https://github.com/Netflix/dgs-framework/blob/master/graphql-dgs-client/src/main/kotlin/com/netflix/graphql/dgs/client/WebClientGraphQLClient.kt#L79

Thanks a lot Paul, this worked like a charm. Appreciate your help. I am using DGS version 4.1.0 and cannot upgrade right now, therefore I had to make it work with DefaultGraphQLClient.
